subsys/bluetooth/audio/pacs.c

```diff
@@ -597,86 +555,152 @@ void bt_pacs_capabilities_changed(enum bt_audio_dir dir)
 {
-	struct k_work_delayable *work;
+	struct bt_pacs_context context = {
+		.snk = sys_cpu_to_le16(snk_available_contexts),
+		.src = sys_cpu_to_le16(src_available_contexts),
+	};
+	int err;
 
-	work = bt_pacs_get_loc_work(dir);
-	if (!work) {
-		return -EINVAL;
+	err = bt_gatt_notify_uuid(NULL, BT_UUID_PACS_AVAILABLE_CONTEXT, pacs_svc.attrs,
+				  &context, sizeof(context));
+	if (err != 0 && err != -ENOTCONN) {
+		BT_WARN("Available Audio Contexts notify failed: %d", err);
 	}
+}
 
-	k_work_reschedule(work, PAC_NOTIFY_TIMEOUT);
+bool bt_pacs_context_available(enum bt_audio_dir dir, uint16_t context)
+{
+	if (dir == BT_AUDIO_DIR_SOURCE) {
+		return (context & src_available_contexts) == context;
+	}
 
-	return 0;
+	if (dir == BT_AUDIO_DIR_SINK) {
+		return (context & snk_available_contexts) == context;
+	}
+
+	return false;
 }
-#endif /* CONFIG_BT_PAC_SNK_LOC || CONFIG_BT_PAC_SRC_LOC */
 
-static void available_contexts_notify(struct k_work *work)
+static sys_slist_t *bt_audio_capability_get(enum bt_audio_dir dir)
 {
-	struct bt_pacs_context context;
-	int err;
+	switch (dir) {
+	case BT_AUDIO_DIR_SINK:
+		return &snks;
+	case BT_AUDIO_DIR_SOURCE:
+		return &srcs;
+	}
+
+	return NULL;
+}
 
-	err = available_contexts_get(NULL, &context);
-	if (err) {
-		BT_DBG("get_available_contexts returned %d, won't notify", err);
+void bt_audio_foreach_capability(enum bt_audio_dir dir, bt_audio_foreach_capability_func_t func,
+				 void *user_data)
+{
+	struct bt_audio_capability *cap;
+	sys_slist_t *lst;
+
+	CHECKIF(func == NULL) {
+		BT_ERR("func is NULL");
 		return;
 	}
 
-	err = bt_gatt_notify_uuid(NULL, BT_UUID_PACS_AVAILABLE_CONTEXT, pacs_svc.attrs,
-				  &context, sizeof(context));
-	if (err != 0 && err != -ENOTCONN) {
-		BT_WARN("Available Audio Contexts notify failed: %d", err);
+	lst = bt_audio_capability_get(dir);
+	if (!lst) {
+		return;
 	}
-}
 
-static K_WORK_DELAYABLE_DEFINE(available_contexts_work, available_contexts_notify);
+	SYS_SLIST_FOR_EACH_CONTAINER(lst, cap, _node) {
+		if (!func(cap, user_data)) {
+			break;
+		}
+	}
+}
 
-int bt_pacs_available_contexts_changed(void)
+/* Register Audio Capability */
+int bt_audio_capability_register(enum bt_audio_dir dir, struct bt_audio_capability *cap)
 {
-	int err;
+	sys_slist_t *lst;
 
-	err = k_work_reschedule(&available_contexts_work, PAC_NOTIFY_TIMEOUT);
-	if (err < 0) {
-		return err;
+	if (!cap || !cap->codec) {
+		return -EINVAL;
 	}
 
+	lst = bt_audio_capability_get(dir);
+	if (!lst) {
+		return -EINVAL;
+	}
+
+	BT_DBG("cap %p dir 0x%02x codec 0x%02x codec cid 0x%04x "
+	       "codec vid 0x%04x", cap, dir, cap->codec->id,
+	       cap->codec->cid, cap->codec->vid);
+
+	sys_slist_append(lst, &cap->_node);
+
+	bt_pacs_capabilities_changed(dir);
+
 	return 0;
 }
 
-int bt_audio_pacs_register_cb(const struct bt_audio_pacs_cb *cb)
+/* Unregister Audio Capability */
+int bt_audio_capability_unregister(enum bt_audio_dir dir, struct bt_audio_capability *cap)
```

Comment:
A general comment: I agree with the general point of this PR, which is to merge the two files (pacs.c and capabilities.c). However with this direct merge, I think we have some namespace challenges here, as we now have both `bt_pacs` and `bt_audio_capability` in the `pacs.c` file. 

Not sure if this is strictly disallowed in Zephyr, but it certainly isn't particularly great to mix namespaces like that in a single file. 

I'd suggest to use either `bt_audio_capability` or `bt_pacs`, instead of this mix. What do you think?

Reply:
FWIW bt_pacs is much shorter and may suits better as prefix :)

Reply:
> FWIW bt_pacs is much shorter and may suits better as prefix :)

To some extent, yes :) `bt_pacs` does refer to an API for a service (which by now kind of is the case). I am not against that, but it is different than what was originally intended for `bt_audio_capability`, as that was meant to serve as more than just a PACS API I believe (@Vudentz would know, but not sure if he follows this anymore). 

I am OK with using either `bt_audio_capability` or `bt_pacs`. 

Unlike most other functionality, we cannot "hide" the PACS behind a specific role (e.g. `unicast_server` or `broadcast_sink`), as both `unicast_server` and `broadcast_sink` use PACS. 

We will actually have a similar challenge with the client access to this, as PACS can be read by the BAP Broadcast assistant, but all the access to PACS is currently behind the `bt_audio_unicast_client` API. Maybe we will need to consider using `bt_pacs` and `bt_pacs_client` for this purpose (similar to `bt_vocs` and `bt_vocs_client`), where the service may be used by different roles.